File: 0051-n-queens/0051-n-queens.cpp
class Solution {
public:
vector<vector<string>> solveNQueens(int n)
{
    vector<string> board(n, string(n, '.'));
    vector<vector<string>> ans;
    helper(board, 0, ans);
    return ans;
}
void helper(vector<string> &board, int col, vector<vector<string>> &ans)
{
    int n = board.size();
    // base case..
    if (col == n)
    {
        ans.push_back(board);
        return;
    }

    for (int i = 0; i < n; i++)
    {

        if (issafe(board, i, col))
        {
            board[i][col] = 'Q';
            helper(board, col + 1, ans);
            board[i][col] = '.';
        }
    }
}

bool issafe(vector<string> &board, int row, int col)
{
    int n = board.size();
    int tempRow = row;
    int tempCol = col;
    while (tempRow >= 0 && tempCol >= 0)
    {
        if (board[tempRow][tempCol] == 'Q')
            return false;
        tempRow--;
        tempCol--;
    }
    tempRow = row, tempCol = col;
    while (tempRow < n && tempCol >= 0)
    {
        if (board[tempRow][tempCol] == 'Q')
            return false;
        tempRow++;
        tempCol--;
    }
    tempRow = row;
    tempCol = col;
    while (tempCol >= 0)
    {
        if (board[tempRow][tempCol] == 'Q')
            return false;
        tempCol--;
    }
    return true;
}
};

File: 0051-n-queens/0051-n-queens.java
class Solution {
    public boolean isSafe(List<StringBuilder> board, int i, int j) {
        int n = board.size();
        // check row ...

        for (int c = 0; c < j; c++) {
            if (board.get(i).charAt(c) == 'Q')
                return false;
        }
        // upper-left diagonal
        for (int row = i - 1, col = j - 1; row >= 0 && col >= 0; row--, col--) {
            if (board.get(row).charAt(col) == 'Q')
                return false;
        }
        // lower-left diagonal
        for (int row = i + 1, col = j - 1; row < n && col >= 0; row++, col--) {
            if (board.get(row).charAt(col) == 'Q')
                return false;
        }
        return true;
    }

    public void helper(List<StringBuilder> board, List<List<String>> ans, int col) {
        int n = board.size();
        if (n == col) {
            List<String> temp = new ArrayList<>();
            for (StringBuilder s : board) {
                temp.add(s.toString());
            }
            ans.add(new ArrayList<>(temp));
            return;
        }
        for (int i = 0; i < n; i++) {
            if (isSafe(board, i, col)) {
                board.get(i).setCharAt(col, 'Q');
                helper(board, ans, col + 1);
                board.get(i).setCharAt(col, '.');
            }
        }
        return;
    }

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> ans = new ArrayList<>();
        List<StringBuilder> board = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            StringBuilder s = new StringBuilder();
            for (int c = 0; c < n; c++)
                s.append('.');
            board.add(s);
        }

        helper(board, ans, 0);
        return ans;
    }
}